File: src/wiki-manager.ts
import * as fs from "fs";
import * as path from "path";
import { generateClaudeMd } from "./templates/claude-md.js";
import { generateReadmeMd } from "./templates/readme-md.js";
import { walkMarkdownFiles } from "./utils.js";

export interface WikiConfig {
  wikiRoot: string; // wiki 子目录名，默认 ""
  pageTypes: string[];
}

const DEFAULT_PAGE_TYPES = [
  "summaries",
  "concepts",
  "entities",
  "methods",
  "comparisons",
  "analysis",
  "indexes",
];

const RAW_SUBDIRS = ["tech", "work", "reading", "general", "assets"];

export class WikiManager {
  private vaultPath: string;
  private wikiRoot: string;

  constructor(vaultPath: string, wikiRoot?: string) {
    this.vaultPath = path.resolve(vaultPath);
    this.wikiRoot = wikiRoot ?? "";
  }

  private resolveWiki(...segments: string[]): string {
    const base = this.wikiRoot
      ? path.join(this.vaultPath, this.wikiRoot)
      : this.vaultPath;
    return path.join(base, ...segments);
  }

  private resolveRaw(...segments: string[]): string {
    return this.resolveWiki("raw", ...segments);
  }

  /**
   * 检测 wiki 是否已初始化
   */
  detect(): {
    initialized: boolean;
    wikiRoot: string;
    rawDirs: string[];
    wikiDirs: string[];
  } {
    const rawDir = this.resolveRaw();
    const wikiDir = this.resolveWiki("wiki");

    const rawExists = fs.existsSync(rawDir);
    const wikiExists = fs.existsSync(wikiDir);

    const rawDirs = rawExists
      ? fs.readdirSync(rawDir).filter((d) =>
          fs.statSync(path.join(rawDir, d)).isDirectory()
        )
      : [];
    const wikiDirs = wikiExists
      ? fs.readdirSync(wikiDir).filter((d) =>
          fs.statSync(path.join(wikiDir, d)).isDirectory()
        )
      : [];

    return {
      initialized: rawExists && wikiExists,
      wikiRoot: this.wikiRoot,
      rawDirs,
      wikiDirs,
    };
  }

  /**
   * 初始化 wiki 目录结构
   */
  init(options?: { pageTypes?: string[] }): {
    created: string[];
    files: string[];
  } {
    const pageTypes = options?.pageTypes ?? DEFAULT_PAGE_TYPES;
    const created: string[] = [];
    const files: string[] = [];

    // 创建 raw 子目录
    for (const sub of RAW_SUBDIRS) {
      const dir = this.resolveRaw(sub);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
        created.push(`raw/${sub}`);
      }
    }

    // 创建 wiki 子目录
    for (const pt of pageTypes) {
      const dir = this.resolveWiki("wiki", pt);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
        created.push(`wiki/${pt}`);
      }
    }

    // 生成 CLAUDE.md
    const claudeMdPath = this.resolveWiki("CLAUDE.md");
    const vaultName = path.basename(this.vaultPath);
    const claudeMdContent = generateClaudeMd({
      vaultName,
      wikiRoot: this.wikiRoot ? `${this.wikiRoot}/` : "",
      pageTypes,
    });
    fs.writeFileSync(claudeMdPath, claudeMdContent, "utf-8");
    files.push("CLAUDE.md");

    // 生成 README.md
    const readmePath = this.resolveWiki("README.md");
    const readmeContent = generateReadmeMd({ vaultName, wikiRoot: this.wikiRoot });
    fs.writeFileSync(readmePath, readmeContent, "utf-8");
    files.push("README.md");

    // 创建索引
    const indexPath = this.resolveWiki("wiki", "indexes", "index.md");
    if (!fs.existsSync(indexPath)) {
      const indexContent = `---\ntype: index\ncreated: ${new Date().toISOString().slice(0, 10)}\ntags: [wiki, index]\n---\n\n# Wiki 索引\n\n> 所有 wiki 页面的目录\n\n## 按类型\n${pageTypes.map((t) => `- [[wiki/${t}/]]`).join("\n")}\n`;
      fs.mkdirSync(path.dirname(indexPath), { recursive: true });
      fs.writeFileSync(indexPath, indexContent, "utf-8");
      files.push("wiki/indexes/index.md");
    }

    // 创建日志
    const logPath = this.resolveWiki("wiki", "indexes", "log.md");
    if (!fs.existsSync(logPath)) {
      const logContent = `---\ntype: index\ntags: [wiki, log]\n---\n\n# 操作日志\n\n## ${new Date().toISOString().slice(0, 10)}\n\n- Wiki 初始化完成\n- 创建目录: ${created.join(", ")}\n`;
      fs.writeFileSync(logPath, logContent, "utf-8");
      files.push("wiki/indexes/log.md");
    }

    return { created, files };
  }

  /**
   * 获取 wiki 状态
   */
  status(): {
    initialized: boolean;
    totalPages: number;
    pageCounts: Record<string, number>;
    rawFiles: string[];
    ingestedSources: string[];
    orphanPages: string[];
  } {
    const detection = this.detect();
    if (!detection.initialized) {
      return {
        initialized: false,
        totalPages: 0,
        pageCounts: {},
        rawFiles: [],
        ingestedSources: [],
        orphanPages: [],
      };
    }

    // 统计各类型页面数
    const pageCounts: Record<string, number> = {};
    let totalPages = 0;
    for (const pt of DEFAULT_PAGE_TYPES) {
      const dir = this.resolveWiki("wiki", pt);
      const files = fs.existsSync(dir)
        ? fs.readdirSync(dir).filter((f) => f.endsWith(".md"))
        : [];
      pageCounts[pt] = files.length;
      totalPages += files.length;
    }

    // 列出 raw 文件
    const rawDir = this.resolveRaw();
    const rawFiles = fs.existsSync(rawDir)
      ? walkMarkdownFiles(rawDir, rawDir)
      : [];

    // 检查已摄入的源（通过 summaries 的 frontmatter sources 字段）
    const summariesDir = this.resolveWiki("wiki", "summaries");
    const ingestedSources: string[] = [];
    if (fs.existsSync(summariesDir)) {
      const summaryFiles = fs.readdirSync(summariesDir).filter((f) => f.endsWith(".md"));
      for (const sf of summaryFiles) {
        const content = fs.readFileSync(path.join(summariesDir, sf), "utf-8");
        const sourcesMatch = content.match(/sources:\s*\[(.+?)\]/);
        if (sourcesMatch) {
          ingestedSources.push(sf.replace(/\.md$/, ""));
        }
      }
    }

    return {
      initialized: true,
      totalPages,
      pageCounts,
      rawFiles,
      ingestedSources,
      orphanPages: [], // 将由 linter 填充
    };
  }

  /**
   * 列出 raw 目录中的源文件及摄入状态
   */
  listSources(folder?: string): Array<{
    path: string;
    ingested: boolean;
    size: number;
  }> {
    const targetDir = folder
      ? this.resolveRaw(folder)
      : this.resolveRaw();

    if (!fs.existsSync(targetDir)) return [];

    const rawBase = this.resolveRaw();
    const files = walkMarkdownFiles(targetDir, rawBase);
    const ingested = new Set(this.status().ingestedSources);

    return files.map((f) => {
      const absPath = path.join(rawBase, f);
      const stat = fs.statSync(absPath);
      const name = path.basename(f, ".md");
      return {
        path: f,
        ingested: ingested.has(name),
        size: stat.size,
      };
    });
  }

  /**
   * 追加操作日志
   */
  appendLog(entry: string): void {
    const logPath = this.resolveWiki("wiki", "indexes", "log.md");
    if (!fs.existsSync(logPath)) return;

    const date = new Date().toISOString().slice(0, 10);
    const content = fs.readFileSync(logPath, "utf-8");

    // 检查今天的日志段是否存在
    if (content.includes(`## ${date}`)) {
      // 追加到今天的段
      const updated = content.replace(
        `## ${date}\n`,
        `## ${date}\n\n- ${entry}\n`,
      );
      fs.writeFileSync(logPath, updated, "utf-8");
    } else {
      // 新增今天的段
      fs.writeFileSync(
        logPath,
        content.trimEnd() + `\n\n## ${date}\n\n- ${entry}\n`,
        "utf-8",
      );
    }
  }

  /**
   * 更新索引页面
   */
  updateIndex(newPages: Array<{ type: string; path: string; title: string }>): void {
    const indexPath = this.resolveWiki("wiki", "indexes", "index.md");
    if (!fs.existsSync(indexPath)) return;

    let content = fs.readFileSync(indexPath, "utf-8");

    for (const page of newPages) {
      const line = `- [[${page.path.replace(/\.md$/, "")}|${page.title}]]`;
      if (!content.includes(line)) {
        // 在对应类型的 section 下追加
        const sectionHeader = `### ${page.type}`;
        if (content.includes(sectionHeader)) {
          content = content.replace(
            sectionHeader + "\n",
            sectionHeader + "\n" + line + "\n",
          );
        } else {
          // 追加新类型 section
          content = content.trimEnd() + `\n\n### ${page.type}\n\n${line}\n`;
        }
      }
    }

    fs.writeFileSync(indexPath, content, "utf-8");
  }
}

File: src/obsidian_wiki_langgraph/core/wiki_manager.py
"""
Wiki Manager - manages wiki directory structure and metadata.
Ported from obsidian-llm-wiki src/wiki-manager.ts
"""

from __future__ import annotations

import re
from datetime import date
from pathlib import Path
from typing import Any

from ..config import DEFAULT_PAGE_TYPES, RAW_SUBDIRS
from ..templates.claude_md import generate_claude_md
from ..templates.readme_md import generate_readme_md
from .utils import walk_markdown_files


class WikiManager:
    """Manages the wiki directory structure, indexes, and logs."""

    def __init__(self, vault_path: str | Path, wiki_root: str | None = None) -> None:
        self.vault_path: Path = Path(vault_path).resolve()
        self.wiki_root: str = wiki_root if wiki_root is not None else ""

    # --- Private path helpers ------------------------------------------------

    def _resolve_wiki(self, *segments: str) -> Path:
        """Resolve a path relative to the wiki root."""
        base = self.vault_path / self.wiki_root if self.wiki_root else self.vault_path
        return base.joinpath(*segments)

    def _resolve_raw(self, *segments: str) -> Path:
        """Resolve a path relative to raw/ under the wiki root."""
        return self._resolve_wiki("raw", *segments)

    # --- Public API ----------------------------------------------------------

    def detect(self) -> dict:
        """Check whether the wiki has been initialised.

        Returns
        -------
        dict
            Keys: initialized, wiki_root, raw_dirs, wiki_dirs.
        """
        raw_dir = self._resolve_raw()
        wiki_dir = self._resolve_wiki("wiki")

        raw_exists = raw_dir.exists()
        wiki_exists = wiki_dir.exists()

        raw_dirs: list[str] = (
            [d.name for d in raw_dir.iterdir() if d.is_dir()]
            if raw_exists
            else []
        )
        wiki_dirs: list[str] = (
            [d.name for d in wiki_dir.iterdir() if d.is_dir()]
            if wiki_exists
            else []
        )

        return {
            "initialized": raw_exists and wiki_exists,
            "wiki_root": self.wiki_root,
            "raw_dirs": raw_dirs,
            "wiki_dirs": wiki_dirs,
        }

    def init(self, page_types: list[str] | None = None) -> dict:
        """Initialise the wiki directory structure.

        Parameters
        ----------
        page_types : list[str] | None
            Wiki page-type sub-directories to create under ``wiki/``.
            Defaults to :data:`DEFAULT_PAGE_TYPES`.

        Returns
        -------
        dict
            Keys: created (list[str]), files (list[str]).
        """
        if page_types is None:
            page_types = list(DEFAULT_PAGE_TYPES)

        created: list[str] = []
        files: list[str] = []

        # Create raw sub-directories
        for sub in RAW_SUBDIRS:
            dir_path = self._resolve_raw(sub)
            if not dir_path.exists():
                dir_path.mkdir(parents=True, exist_ok=True)
                created.append(f"raw/{sub}")

        # Create wiki sub-directories
        for pt in page_types:
            dir_path = self._resolve_wiki("wiki", pt)
            if not dir_path.exists():
                dir_path.mkdir(parents=True, exist_ok=True)
                created.append(f"wiki/{pt}")

        # Generate CLAUDE.md
        claude_md_path = self._resolve_wiki("CLAUDE.md")
        vault_name = self.vault_path.name
        claude_md_content = generate_claude_md({
            "vault_name": vault_name,
            "wiki_root": f"{self.wiki_root}/" if self.wiki_root else "",
            "page_types": page_types,
        })
        claude_md_path.write_text(claude_md_content, encoding="utf-8")
        files.append("CLAUDE.md")

        # Generate README.md
        readme_path = self._resolve_wiki("README.md")
        readme_content = generate_readme_md({
            "vault_name": vault_name,
            "wiki_root": self.wiki_root,
        })
        readme_path.write_text(readme_content, encoding="utf-8")
        files.append("README.md")

        # Create index page
        index_path = self._resolve_wiki("wiki", "indexes", "index.md")
        if not index_path.exists():
            today = date.today().isoformat()
            page_type_lines = "\n".join(
                f"- [[wiki/{t}/]]" for t in page_types
            )
            index_content = (
                f"---\n"
                f"type: index\n"
                f"created: {today}\n"
                f"tags: [wiki, index]\n"
                f"---\n\n"
                f"# Wiki 索引\n\n"
                f"> 所有 wiki 页面的目录\n\n"
                f"## 按类型\n"
                f"{page_type_lines}\n"
            )
            index_path.parent.mkdir(parents=True, exist_ok=True)
            index_path.write_text(index_content, encoding="utf-8")
            files.append("wiki/indexes/index.md")

        # Create log page
        log_path = self._resolve_wiki("wiki", "indexes", "log.md")
        if not log_path.exists():
            today = date.today().isoformat()
            log_content = (
                f"---\n"
                f"type: index\n"
                f"tags: [wiki, log]\n"
                f"---\n\n"
                f"# 操作日志\n\n"
                f"## {today}\n\n"
                f"- Wiki 初始化完成\n"
                f"- 创建目录: {', '.join(created)}\n"
            )
            log_path.write_text(log_content, encoding="utf-8")
            files.append("wiki/indexes/log.md")

        return {"created": created, "files": files}

    def status(self) -> dict:
        """Return the current wiki status.

        Returns
        -------
        dict
            Keys: initialized, total_pages, page_counts, raw_files,
            ingested_sources, orphan_pages.
        """
        detection = self.detect()
        if not detection["initialized"]:
            return {
                "initialized": False,
                "total_pages": 0,
                "page_counts": {},
                "raw_files": [],
                "ingested_sources": [],
                "orphan_pages": [],
            }

        # Count pages per type
        page_counts: dict[str, int] = {}
        total_pages = 0
        for pt in DEFAULT_PAGE_TYPES:
            dir_path = self._resolve_wiki("wiki", pt)
            if dir_path.exists():
                md_files = [f for f in dir_path.iterdir() if f.suffix == ".md"]
                count = len(md_files)
            else:
                count = 0
            page_counts[pt] = count
            total_pages += count

        # List raw files
        raw_dir = self._resolve_raw()
        raw_files: list[str] = (
            walk_markdown_files(raw_dir, raw_dir) if raw_dir.exists() else []
        )

        # Check ingested sources (via summaries frontmatter)
        summaries_dir = self._resolve_wiki("wiki", "summaries")
        ingested_sources: list[str] = []
        if summaries_dir.exists():
            for sf in summaries_dir.iterdir():
                if sf.suffix != ".md":
                    continue
                content = sf.read_text(encoding="utf-8")
                if re.search(r"sources:\s*\[.+?\]", content):
                    ingested_sources.append(sf.stem)

        return {
            "initialized": True,
            "total_pages": total_pages,
            "page_counts": page_counts,
            "raw_files": raw_files,
            "ingested_sources": ingested_sources,
            "orphan_pages": [],
        }

    def list_sources(self, folder: str | None = None) -> list[dict]:
        """List raw source files with their ingestion status.

        Parameters
        ----------
        folder : str | None
            Sub-directory of raw/ to list.  Lists all of raw/ when *None*.

        Returns
        -------
        list[dict]
            Each dict has keys: path, ingested, size.
        """
        target_dir = self._resolve_raw(folder) if folder else self._resolve_raw()

        if not target_dir.exists():
            return []

        raw_base = self._resolve_raw()
        files = walk_markdown_files(target_dir, raw_base)
        ingested = set(self.status()["ingested_sources"])

        result: list[dict] = []
        for f in files:
            abs_path = raw_base / f
            stat = abs_path.stat()
            name = Path(f).stem
            result.append({
                "path": f,
                "ingested": name in ingested,
                "size": stat.st_size,
            })
        return result

    def append_log(self, entry: str) -> None:
        """Append an entry to the wiki operation log.

        Parameters
        ----------
        entry : str
            The log message to append.
        """
        log_path = self._resolve_wiki("wiki", "indexes", "log.md")
        if not log_path.exists():
            return

        today = date.today().isoformat()
        content = log_path.read_text(encoding="utf-8")

        # Check whether today's section already exists
        section_header = f"## {today}"
        if section_header in content:
            # Append to today's section
            updated = content.replace(
                f"{section_header}\n",
                f"{section_header}\n\n- {entry}\n",
            )
            log_path.write_text(updated, encoding="utf-8")
        else:
            # Create a new section for today
            log_path.write_text(
                content.rstrip() + f"\n\n{section_header}\n\n- {entry}\n",
                encoding="utf-8",
            )

    def update_index(self, new_pages: list[dict]) -> None:
        """Update the wiki index with new pages.

        Parameters
        ----------
        new_pages : list[dict]
            Each dict must have keys: type, path, title.
        """
        index_path = self._resolve_wiki("wiki", "indexes", "index.md")
        if not index_path.exists():
            return

        content = index_path.read_text(encoding="utf-8")

        for page in new_pages:
            line = f"- [[{page['path'].removesuffix('.md')}|{page['title']}]]"
            if line in content:
                continue

            section_header = f"### {page['type']}"
            if section_header in content:
                content = content.replace(
                    f"{section_header}\n",
                    f"{section_header}\n{line}\n",
                )
            else:
                # Append a new type section
                content = content.rstrip() + f"\n\n{section_header}\n\n{line}\n"

        index_path.write_text(content, encoding="utf-8")
